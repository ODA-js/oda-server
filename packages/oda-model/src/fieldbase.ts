import {
  ModelBase,
  IModelBase,
  ModelBaseInternal,
  ModelBaseInput,
  ModelBaseOutput,
} from './modelbase';
import decapitalize from './lib/decapitalize';
import { merge, get } from 'lodash';
import {
  FieldArgs,
  FieldType,
  AsHash,
  MetaModelType,
  HashToMap,
  MapToHash,
  Nullable,
  assignValue,
  NamedArray,
  ArrayToMap,
} from './types';
import { ElementMetaInfo } from './element';
import { IEntityRef, EntityReference } from './entityreference';

export interface IFieldBase<
  M extends FieldBaseMetaInfo<P>,
  I extends FieldBaseInput<M, P>,
  P extends FieldBasePersistence,
  O extends FieldBaseOutput<M, P>
> extends IModelBase<M, I, O> {
  args: Map<string, FieldArgs>;
  inheritedFrom?: string;
  type?: FieldType;
  required?: boolean;
  indexed: boolean | string | string[];
  identity: boolean | string | string[];
  idKey: IEntityRef;
}

export interface FieldBasePersistence {
  derived: boolean;
  persistent: boolean;
  required: boolean;
  identity: boolean | string | string[];
  indexed: boolean | string | string[];
}

export interface FieldBaseMetaInfo<T extends FieldBasePersistence>
  extends ElementMetaInfo {
  entity: string;
  persistence: T;
  order: number;
}

export interface FieldBaseInternal<
  T extends FieldBaseMetaInfo<P>,
  P extends FieldBasePersistence
> extends ModelBaseInternal<T> {
  args: Map<string, FieldArgs>;
  inheritedFrom?: string;
  type?: FieldType;
  idKey: IEntityRef;
}

export interface FieldBaseInput<
  T extends FieldBaseMetaInfo<P>,
  P extends FieldBasePersistence
> extends ModelBaseInput<T> {
  args?: AsHash<FieldArgs> | NamedArray<FieldArgs>;
  inheritedFrom?: string;
  derived?: boolean;
  persistent?: boolean;
  entity?: string;
  type?: FieldType;
  order?: number;
  required?: boolean;
  identity?: boolean | string | string[];
  indexed?: boolean | string | string[];
}

export interface FieldBaseOutput<
  T extends FieldBaseMetaInfo<P>,
  P extends FieldBasePersistence
> extends ModelBaseOutput<T> {
  inheritedFrom?: string;
  derived?: boolean;
  persistent?: boolean;
  entity?: string;
  type?: FieldType;
  order?: number;
  required?: boolean;
  identity?: boolean | string | string[];
  indexed?: boolean | string | string[];
  args: NamedArray<FieldArgs>;
}

const defaultMetaInfo = {};
const defaultInternal = {};
const defaultInput = {};

export abstract class FieldBase<
  T extends FieldBaseMetaInfo<P>,
  I extends FieldBaseInput<T, P>,
  S extends FieldBaseInternal<T, P>,
  P extends FieldBasePersistence,
  O extends FieldBaseOutput<T, P>
> extends ModelBase<T, I, S, O> implements IFieldBase<T, I, P, O> {
  public modelType: MetaModelType = 'field-base';
  public get idKey(): IEntityRef {
    return this.$obj.idKey;
  }

  get type(): FieldType | undefined {
    return this.$obj.type;
  }

  get inheritedFrom(): string | undefined {
    return this.$obj.inheritedFrom;
  }

  get args(): Map<string, FieldArgs> {
    return this.$obj.args;
  }

  get order(): number {
    return this.metadata_.order;
  }

  get derived(): boolean {
    return get(this.metadata_, 'persistence.derived', false);
  }

  get persistent(): boolean {
    return get(this.metadata_, 'persistence.persistent', false);
  }

  get identity(): boolean | string | string[] {
    return get(this.metadata_, 'persistence.identity', false);
  }

  get required(): boolean {
    return get(this.metadata_, 'persistence.required', false);
  }

  get indexed(): boolean | string | string[] {
    return get(this.metadata_, 'persistence.indexed', false);
  }

  constructor(inp: I) {
    super(merge({}, defaultInput, inp));
    this.metadata_ = merge({}, defaultMetaInfo, this.metadata_);
    this.$obj = merge({}, defaultInternal, this.$obj);
  }

  public updateWith(input: Nullable<I>) {
    super.updateWith(input);

    assignValue<S, I, string>({
      src: this.$obj,
      input,
      field: 'name',
      effect: (src, value) => {
        src.name = decapitalize(value);
      },
      required: true,
    });

    assignValue({
      src: this.$obj,
      input,
      field: 'inheritedFrom',
    });

    assignValue<S, I, AsHash<FieldArgs> | NamedArray<FieldArgs>>({
      src: this.$obj,
      input,
      field: 'args',
      effect: (src, value) =>
        (src.args = Array.isArray(value)
          ? ArrayToMap(value)
          : HashToMap(value)),
      setDefault: src => (src.args = new Map()),
    });

    assignValue({
      src: this.metadata_,
      input,
      field: 'entity',
      effect: (src, value) => (src.entity = value),
    });

    assignValue({
      src: this.metadata_,
      input,
      field: 'order',
      effect: (src, value) => (src.order = value),
    });

    assignValue({
      src: this.metadata_,
      input,
      inputField: 'derived',
      effect: (src, value) => (src.persistence.derived, value),
    });

    assignValue({
      src: this.metadata_,
      input,
      inputField: 'persistent',
      effect: (src, value) => (src.persistence.persistent = value),
    });

    assignValue<T, I, boolean>({
      src: this.metadata_,
      input,
      inputField: 'identity',
      effect: (_, value) => {
        this.metadata_.persistence.identity = value;
        if (value) {
          this.metadata_.persistence.required = true;
          this.metadata_.persistence.indexed = true;
          this.$obj.idKey = new EntityReference({
            entity: this.metadata_.entity,
            field: this.$obj.name,
          });
        }
      },
    });

    assignValue<T, I, boolean>({
      src: this.metadata_,
      input,
      inputField: 'indexed',
      effect: (src, value) => (src.persistence.indexed = value),
    });

    assignValue<T, I, boolean>({
      src: this.metadata_,
      input,
      inputField: 'required',
      effect: (src, value) => (src.persistence.required = value),
    });
  }

  public makeIdentity() {
    this.$obj.idKey = new EntityReference({
      entity: this.$obj.metadata.entity,
      field: this.$obj.name,
      backField: 'id',
    });
    this.metadata_.persistence.indexed = true;
    this.metadata_.persistence.identity = true;
    this.metadata_.persistence.required = true;
  }

  public toObject(): O {
    return merge({}, super.toObject(), {
      entity: this.metadata_.entity,
      derived: this.derived,
      persistent: this.persistent,
      inheritedFrom: this.$obj.inheritedFrom,
      order: this.metadata_.order,
      args: this.$obj.args ? MapToHash(this.$obj.args) : undefined,
      required: this.required,
      indexed: this.indexed,
      identity: this.identity,
    } as Partial<O>);
  }
}
