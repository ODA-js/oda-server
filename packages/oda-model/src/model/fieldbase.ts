import clean from '../lib/json/clean';
import {
  FieldArgs,
  FieldBaseInput,
  FieldBaseStorage,
  MetaModelType,
  FieldType,
} from './interfaces';
import { ModelBase } from './modelbase';
import decapitalize from './../lib/decapitalize';

export class FieldBase extends ModelBase {
  public modelType: MetaModelType = 'field';
  protected $obj: FieldBaseStorage;

  get entity(): string {
    return this.$obj.entity;
  }

  get type(): FieldType {
    return this.$obj.type;
  }

  get inheritedFrom(): string {
    return this.$obj.inheritedFrom;
  }

  get args(): FieldArgs[] {
    return this.$obj.args;
  }

  public updateWith(obj: FieldBaseInput) {
    if (obj) {
      super.updateWith(obj);

      const result = { ...this.$obj };

      this.$obj.name = decapitalize(this.$obj.name);
      this.$obj.name_ = decapitalize(this.$obj.name_);

      let $entity = obj.entity;
      let entity = $entity;

      let args = obj.args;
      let $args = obj.args;

      let $type = obj.type;
      let type = $type || 'String';

      result.inheritedFrom = obj.inheritedFrom;

      result.type_ = $type;
      result.type = type;

      result.entity = entity;
      result.entity_ = $entity;

      result.args = args;
      result.args_ = $args;

      this.$obj = result;
    }
  }

  // it get fixed object
  public toObject() {
    let props = this.$obj;
    let res = super.toObject();
    return clean({
      ...res,
      entity: props.entity || props.entity_,
      type: props.type_,
      inheritedFrom: props.inheritedFrom,
      args: props.args || props.args_,
    });
  }

  // it get clean object with no default values
  public toJSON() {
    let props = this.$obj;
    let res = super.toJSON();
    return clean({
      ...res,
      type: props.type_,
      inheritedFrom: props.inheritedFrom,
      args: props.args_,
    });
  }
}
