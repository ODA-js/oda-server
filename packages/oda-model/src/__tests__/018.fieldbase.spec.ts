import 'jest';
import { FieldBase } from '../fieldbase';

describe('FieldBase', () => {
  it('default', () => {
    const res = new FieldBase({
      name: 'FieldName',
      derived: true,
      args: [{ name: 'demoARG' }],
      inheritedFrom: '_B',
      persistent: true,
      entity: 'B',
      order: 1,
      required: true,
      identity: true,
      indexed: true,
      type: 'SomString',
    });
    expect(res.name).toBe('fieldName');
    expect(res.toObject()).toMatchSnapshot('toObject clean');
    expect(res).toMatchSnapshot('internal');
  });
});
