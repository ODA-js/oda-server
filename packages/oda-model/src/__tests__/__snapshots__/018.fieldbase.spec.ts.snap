// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldBase default: internal 1`] = `
FieldBase {
  Symbol(internal): Object {
    "args": Map {
      "demoARG" => ObjectTypeField {
        Symbol(internal): Object {
          "kind": "input",
          "multiplicity": "one",
          "name": "demoARG",
          "type": "string",
        },
        Symbol(metadata): Object {
          "required": false,
        },
      },
    },
    "idKey": EntityReference {
      Symbol(internal): Object {
        "entity": "B",
        "field": "fieldName",
      },
      Symbol(metadata): Object {},
    },
    "inheritedFrom": "_B",
    "name": "fieldName",
  },
  Symbol(metadata): Object {
    "acl": Object {
      "read": Array [],
      "update": Array [],
    },
    "entity": "B",
    "order": 1,
    "persistence": Object {
      "derived": true,
      "identity": true,
      "indexed": true,
      "persistent": true,
      "required": true,
    },
  },
}
`;

exports[`FieldBase default: toObject clean 1`] = `
Object {
  "args": Object {
    "demoARG": Object {
      "description": undefined,
      "metadata": Object {
        "required": false,
      },
      "multiplicity": "one",
      "name": "demoARG",
      "title": undefined,
      "type": "string",
    },
  },
  "description": undefined,
  "inheritedFrom": "_B",
  "metadata": Object {
    "acl": Object {
      "read": Array [],
      "update": Array [],
    },
    "entity": "B",
    "order": 1,
    "persistence": Object {
      "derived": true,
      "identity": true,
      "indexed": true,
      "persistent": true,
      "required": true,
    },
  },
  "name": "fieldName",
  "title": undefined,
}
`;

exports[`FieldBase identity make indexed: identity 1`] = `
FieldBase {
  Symbol(internal): Object {
    "args": Map {},
    "idKey": EntityReference {
      Symbol(internal): Object {
        "entity": "FAKE",
        "field": "field",
      },
      Symbol(metadata): Object {},
    },
    "name": "field",
  },
  Symbol(metadata): Object {
    "acl": Object {
      "read": Array [],
      "update": Array [],
    },
    "persistence": Object {
      "derived": false,
      "identity": true,
      "indexed": true,
      "persistent": true,
      "required": true,
    },
  },
}
`;
